pyo3-macros-backend/src/params.rs

```diff
@@ -161,14 +165,20 @@ pub fn impl_arg_params(
 
     let num_params = positional_parameter_names.len() + keyword_only_parameters.len();
 
-    let mut option_pos = 0;
+    let mut option_pos = 0usize;
     let param_conversion = spec
         .signature
         .arguments
         .iter()
         .enumerate()
         .map(|(i, arg)| {
-            impl_arg_param(arg, i, &mut option_pos, &args_array, holders, ctx).map(|tokens| {
+            let from_py_with = syn::Ident::new(&format!("from_py_with_{}", i), Span::call_site());
+            let arg_value = quote!(#args_array[#option_pos].as_deref());
+            if arg.is_regular() {
+                option_pos += 1;
+            }
```

Comment:
This is a little unfortunate to have the coupling between incrementing the option and `impl_arg_param`. I guess that having the enum, then `impl_arg_param` could be used just for "Regular" variants and then other args like `py` or `cancel_handle` could be their own branches.

Maybe it's better for me to try to implement what I mean than add a drive by comment for that, though?

Reply:
Feel free to continue/add something here if you have something in mind. I wasn't really satisfied by how this turned out either (which is why I left it as draft for now)

Reply:
I have played around a bit with your enum idea over [here](https://github.com/Icxolu/pyo3/commit/d1e270c94eaaae9eaf57249036c042586879460b). Maybe that goes into the direction that you hand in mind? That does definetly decouple the different branches from each other, but it's a bit more involved and not really relevant to the `#[setter]` rework here. So maybe you are right to leave this as is, and rework `FnArg` in a followup.

Reply:
That definitely looks like the idea I had in mind. Does it feel like a step in the right direction as you're working on it?

Reply:
It does feel nicer in my opinion. Having an enum over a bunch of booleans not only makes it much clearer what the branches are, but also make a lot of weird combinations (like optional py/varargs, required kwargs, ...) unrepresentable, which is always good. This also moves the error checking of those cases to construction/parsing side instead of usage side, which for me feels a more appropriate place for them. In some places the matching is bit more verbose, but maybe there could be some helper methods for those (or maybe its even fine like it is)

Reply:
Great! In which case let's merge this one now so that it's possible to start reviewing the follow up :)